src/win/inject.c

```diff
@@ -12,42 +12,54 @@ void injectProcess(HANDLE proc) {
 	FARPROC addr;
 	LPVOID arg;
 	char dll[PATH_MAX];
-	char *ext = 0;
+	int ndll;
 	DWORD rc;
+	static int injecting;
 	extern IMAGE_DOS_HEADER __ImageBase;
 	ASSERT(proc);
+
 	memset(dll, 0, sizeof(dll));
-	CHK(GetModuleFileNameA((HMODULE)&__ImageBase, dll, sizeof(dll)));
-	if (!ext)
-		ext = strstr(dll, ".exe");
-	if (!ext)
-		ext = strstr(dll, ".dll");
-	if (!ext)
-		ext = dll + strlen(dll);
+	CHK(ndll = GetModuleFileNameA((HMODULE)&__ImageBase, dll, sizeof(dll)));
+
+	// dll is one of mypath\fsatrace.exe, mypath\fsatrace64.dll or mypath\fsatrace32.dll
 	CHK(IsWow64Process(proc, &is32));
+	if (strcmp(&dll[ndll-4], ".exe") == 0)
+		ndll -= 4; // .exe
+	else
+		ndll -= 6; // 32.dll or 64.dll
+	memcpy(&dll[ndll], is32 ? "32.dll" : "64.dll", 6);
+
 	CHK(0 != (arg = VirtualAllocEx(proc, 0, strlen(dll) + 1,
 				       MEM_RESERVE | MEM_COMMIT, PAGE_READWRITE)));
-	if (strcmp(ext, ".dll"))
-		memcpy(ext, is32 ? "32.dll" : "64.dll", 6);
-	if (is32) {
+
+	if (is32 && !IS32) {
 		STARTUPINFO si;
 		PROCESS_INFORMATION pi;
 		const char * helpername = "fsatracehelper.exe";
 		char helper[PATH_MAX];
 		char * p;
+
+		// On 32bit Windows when we spawn fsatracehelper it reenters
+		// and tries to run itself again, if we detect this we abort
+		// since we don't need to trace the command fsatracehelper itself
+		if (injecting) return;
```

Comment:
How does this prevent that? Isn't `injecting` in our own address space? 
Perhaps it could be detected via the module filename?

Reply:
Matching on the module filename doesn't work because it ends up being `cmd` that is spawned in newer Windows versions (I suspect that's the change that broke everything). That was my first attempt.

The rough sequence of events is we see 32 bit access, we therefore spawn fsatracehelper, but what Windows does behind the scenes is spawn cmd, which we try to inject into, discover it's 32bit, so spawn fsatracehelper and thus loop forever. This test stops that. I don't quite understand why it's in the right address space, but I think it's plausible that we are spawning cmd in our own address space as a consequence of crossing over to 32bit.

Reply:
> Matching on the module filename doesn't work because it ends up being `cmd` that is spawned in newer Windows versions

That could be serious for my own build system. I'm trying to find a reference explaining that behaviour and/or which windows version starts behaving that way. Do you have any pointer?

Reply:
I discovered this by printf debugging. I have no idea when this started happening, but it definitely happened on my desktop on the latest Windows. I was running Windows Defender, so it's possible it was that, although that's almost a default configuration now, so is almost considered Windows behaviour. I didn't see `cmd` appear in process explorer, so it was all a little odd.

Reply:
This is a bit unclear. Perhaps this nesting is already handled in a less obscure way by the `is32 && !IS32)` condition you added? Can you try removing the `if (injecting) return;` part and see if it passes the test suite?